src/riak_kv_ddl_compiler.erl

```diff
@@ -0,0 +1,1458 @@
+
+validate_fields([], []) ->
+    [];
+validate_fields([], Acc) ->
+    lists:flatten(lists:reverse(Acc));
+validate_fields([#riak_field_v1{name     = N,
+				position = P,
+				type     = Ty,
+				optional = O} | T], Acc)
+  when is_list(N),             % guard 1
+       is_integer(P)           % guard 2
+       andalso P > 0,
+       Ty == binary            % guard 3
+       orelse Ty =:= integer
+       orelse Ty =:= float
+       orelse Ty =:= boolean
+       orelse Ty =:= set
+       orelse Ty =:= timestamp
+       orelse Ty =:= any,
+       is_boolean(O) ->        % guard 4
+    validate_fields(T, Acc);
+validate_fields([#riak_field_v1{name     = N,
+				position = P,
+				type     = {map, MapFields},
+				optional = false} | T], Acc)
+  when is_list(N),   % guard 1
+       is_integer(P) % guard 2
+       andalso P > 0
+       ->
+    NewAcc = case validate_fields(MapFields, []) of
+		 [] -> Acc;
+		 A  -> [A | Acc]
+	     end,
+    validate_fields(T, NewAcc);
+validate_fields([#riak_field_v1{name     = N,
+				position = P,
+				type     = {map, _MapFields},
+				optional = true} = H | T], Acc)
+  when is_list(N),   % guard 1
+       is_integer(P) % guard 2
+       andalso P > 0
+       ->
+    validate_fields(T, [{maps_cant_be_optional, H} | Acc]);
+validate_fields([H | T], Acc) ->
+    validate_fields(T, [{invalid_field, H} | Acc]).
+
+%%
+%% funs to compile the DDL to its helper module AST
+%%
+
+-spec compile(#ddl_v1{}, string(), boolean()) ->
+		     {module, ast()} | {'error', tuple()}.
+compile(#ddl_v1{} = DDL, OutputDir, HasDebugOutput) ->
+    #ddl_v1{bucket        = Bucket,
+	    fields        = Fields} = DDL,
+    {ModName, Attrs, LineNo} = make_attrs(Bucket, ?LINENOSTART),
+    {VFns, LineNo2} = build_validn_fns([Fields], LineNo, ?POSNOSTART, [], []),
+    {ExtractFn, LineNo3} = build_extract_fn([Fields],  LineNo2, []),
+    {GetTypeFn, LineNo4} = build_get_type_fn([Fields], LineNo3, []),
+    {IsValidFn, LineNo5} = build_is_valid_fn([Fields], LineNo4, []),
+    AST = Attrs
+	++ VFns
+	++ [ExtractFn]
+	++ [GetTypeFn]
+	++ [IsValidFn]
+	++ [{eof, LineNo5}],
+    if HasDebugOutput ->
+	    ASTFileName = filename:join([OutputDir, ModName]) ++ ".ast",
+	    write_file(ASTFileName, io_lib:format("~p", [AST]));
+       el/=se ->
+	    ok
+    end,
+    Lint = erl_lint:module(AST),
+    case Lint of
+ 	{ok, []} ->
+     	    if
+     		HasDebugOutput ->
+		    SrcFileName = filename:join([OutputDir, ModName]) ++ ".erl",
+		    write_src(AST, DDL, SrcFileName);
+     		el/=se ->
+		    ok
+     	    end,
+     	    _Module = {ModName, AST};
+     	Other  -> exit(Other)
+    end.
+
+write_src(AST, DDL, SrcFileName) ->
+    AST2 = filter_ast(AST, []),
+    Syntax = erl_syntax:form_list(AST2),
+    Header = io_lib:format("%%% Generated Module, do NOT edit~n~n" ++
+			       "Validates the DDL~n~n" ++
+			       "Bucket        : ~s~n" ++
+			       "Fields        : ~p~n" ++
+			       "Partition_Key : ~p~n~n",
+			   [
```

Comment:
This has new lines (~n) but each line does not start with %%%, will that compile?